### ISSUE TITLE
Keras_to_tpu_model: validation_split parameter does not remove validation samples from training set

### ISSUE THREAD

**Issue by mike-holcomb** [tensorflow/tensorflow#25034]
**System information**
- Using unmodified Google Colab TPU runtime

== cat /etc/issue ===============================================
Linux 07e8e3e8ed41 4.14.79+ #1 SMP Wed Dec 19 21:19:13 PST 2018 x86_64 x86_64 x86_64 GNU/Linux
VERSION="18.04.1 LTS (Bionic Beaver)"
VERSION_ID="18.04"
VERSION_CODENAME=bionic

== are we in docker =============================================
Yes

== compiler =====================================================
c++ (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

== uname -a =====================================================
Linux 07e8e3e8ed41 4.14.79+ #1 SMP Wed Dec 19 21:19:13 PST 2018 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
mesh-tensorflow          0.0.5                
msgpack-numpy            0.4.3.2              
numpy                    1.14.6               
protobuf                 3.6.1                
tensorflow               1.12.0               
tensorflow-hub           0.2.0                
tensorflow-metadata      0.9.0                
tensorflow-probability   0.5.0                

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.12.0
tf.GIT_VERSION = v1.12.0-0-ga6d8ffae09
tf.COMPILER_VERSION = v1.12.0-0-ga6d8ffae09
Sanity check: array([1], dtype=int32)

== env ==========================================================
LD_LIBRARY_PATH /usr/local/nvidia/lib:/usr/local/nvidia/lib64
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running.

== cuda libs  ===================================================
/usr/local/cuda-9.2/targets/x86_64-linux/lib/libcudart.so.9.2.148
/usr/local/cuda-9.2/targets/x86_64-linux/lib/libcudart_static.a
/usr/local/cuda-9.2/doc/man/man7/libcudart.7
/usr/local/cuda-9.2/doc/man/man7/libcudart.so.7

**Describe the current behavior**
When using the `tf.contrib.tpu.keras_to_tpu_model` API for training models on the TPU, providing a `validation_split` parameter does not remove the validation samples from the training set.  This causes the reported validation loss/metrics to be artificially high since we are training on samples that should have been withheld.

**Describe the expected behavior**
Specifically, when calling the fit method with a `validation_split` parameter, a proportion of the samples equal to (count of samples) * (validation_split) should be removed from the training data.

**Code to reproduce the issue**
A Colab notebook that reproduces the issue is here:
https://colab.research.google.com/drive/13g_eEYBUGw9GR6XVjop4TOQEKEDCJQVG

Specifically, a dummy dataset with 256 samples with validation_split = 1/8 should result in 224 training samples and 32 validation samples, but currently shows 256 training samples and 32 validation samples.

**Other info / logs**
I am very grateful for to everyone that is working on this interface.  This makes the on-ramp to TPU training much more manageable.

**Comment by mike-holcomb** [comment]
@jvishnuvardhan Thanks for your quick response.  This is a bug.  Based on your link to a question about the normal `keras.models.Model`, I believe I see the source of the confusion: my issue relates specifically to the `KerasTPUModel` class in `tf.contrib.tpu.keras_support.py.`  

Currently, the `KerasTPUModel` implementation of the `validation_split` parameter in the `fit()` method violates the key assumption behind out-of-sample testing (i.e., that your training set and your evaluation set are mutually exclusive).  

The `KerasTPUModel._prepare_validation_data()` (called by `fit()` to implement the `validation_split` parameter) performs the proper, mutually exclusive split but only modifies a local training set variable. The `_prepare_validation_data()` returns only the new validation subset to the caller and leaves the caller's training set unmodified.  As such, `fit()` uses the original, full data set for training while incorrectly calculating a the validation metrics on the last _validation_split_ % of the training set.

My Colab example and pull request have more details about the issue within the context of `KerasTPUModel`.

As a workaround, I am creating my own training/validation splits and then supplying the validation set via the `validation_data` parameter.

**Comment by AysunTE** [comment]
I have the same problem. I have got different training sets with/without TPU.
